### PR TITLE
Excel range import

### PR DESCRIPTION
**Pull request by naterush**
# Description

Implements PR 1 for here: https://www.notion.so/trymito/Excel-Range-Import-b95bd6054a3a4d62851f028632d72242

# Testing

See tests. Suggest any I should add. This is backend only currently!

# Documentation

Nope.

### CODE REVIEW

File: mitosheet/mitosheet/step_performers/excel_range_import.py [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+        post_state = prev_state.copy() 
+
+        pandas_start_time = perf_counter()
+
+        for df_name, range_import in range_imports.items():
+            ((start_row_index, start_col_index), (end_row_index, end_col_index)) = get_row_and_col_indexes_from_range(range_import['range'])
+            nrows = end_row_index - start_row_index
+            usecols = get_column_from_column_index(start_col_index) + ':' + get_column_from_column_index(end_col_index)
+
+            print("USING", usecols)
```
Comment (naterush) [inline mitosheet/mitosheet/step_performers/excel_range_import.py:52]: Remove

File: mitosheet/mitosheet/step_performers/excel_range_import.py [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+from mitosheet.state import DATAFRAME_SOURCE_IMPORTED, State
+from mitosheet.step_performers.step_performer import StepPerformer
+from mitosheet.step_performers.utils import get_param
+from mitosheet.types import ExcelRangeImportMethod
+from mitosheet.utils import get_valid_dataframe_name
+
+
+class ExcelRangeImportStepPerformer(StepPerformer):
+    """
+    Allows you to excel range import.
```
Comment (naterush) [inline mitosheet/mitosheet/step_performers/excel_range_import.py:26]: Better

File: mitosheet/mitosheet/step_performers/excel_range_import.py [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+                df_name=final_df_name
+            )
+
+        pandas_processing_time = perf_counter() - pandas_start_time
+
+
+        return post_state, {
+            'pandas_processing_time': pandas_processing_time,
+            'result': {
+                # TODO: fill in the result
```
Comment (naterush) [inline mitosheet/mitosheet/step_performers/excel_range_import.py:69]: There is no result!

**Review by naterush — COMMENTED** [review]
Almost good!

File: mitosheet/mitosheet/tests/step_performers/test_excel_range_import.py [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+        [TEST_DF_1]
+    ),
+    (
+        {'dataframe_1': {'type': 'range', 'range': 'A1:B2'}, 'dataframe_2': {'type': 'range', 'range': 'A4:B5'}},
+        [TEST_DF_1, TEST_DF_2]
+    ),
+    (
+        {'a bad dataframe name': {'type': 'range', 'range': 'A1:B2'}, '97 also bad': {'type': 'range', 'range': 'A4:B5'}},
+        [TEST_DF_1, TEST_DF_2]
+    ),
```
Comment (aarondr77) [inline mitosheet/mitosheet/tests/step_performers/test_excel_range_import.py:33]: Some other test ideas 
- [ ] Pass the same dataframe to two range imports 
- [ ] Other range formats, like: A5:A1 and B5:A1
- [ ] Dataframes that have more than just 1 row in them 
- [ ] Check that we handle invalid ranges correctly. For example I don't think we support A:C 
- [ ] On Test_DF_1, what happens if we use the range, A1:G5

File: mitosheet/mitosheet/excel_utils.py [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+    # Add 1 because Mito 0 indexes columns
+    col_index = col_index + 1
+    name = ''
+    while col_index > 0:
+        col_index, r = divmod (col_index - 1, 26)
+        name = chr(r + ord('A')) + name
+    return name
+
+def get_row_and_col_indexes_from_range(range: str) -> Tuple[Tuple[int, int], Tuple[int, int]]:
+    start_cell_address, end_cell_address = range.split(':')
```
Comment (aarondr77) [inline mitosheet/mitosheet/excel_utils.py:27]: We should probably handle errors here or in the helper function if the range does not have a `:` in it.

File: mitosheet/mitosheet/code_chunks/excel_range_import_code_chunk.py [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+
+    def get_display_name(self) -> str:
+        return 'Excel Range Import'
+    
+    def get_description_comment(self) -> str:
+        
+        return f"Imported {len(self.range_imports)} dataframes from {self.sheet_name} in {self.file_name}"
+
+    def get_code(self) -> List[str]:
+        code = []
```
Comment (aarondr77) [inline mitosheet/mitosheet/code_chunks/excel_range_import_code_chunk.py:29]: Until we do the code optimization, this should be `code = ['import pandas as pd']`

**Review by aarondr77 — COMMENTED** [review]
After this review LGTM.

File: mitosheet/dev/metaprogramming/utils/code_utils.py [hunk 7]
```diff
@@ -27,7 +27,7 @@ def get_default_typescript_value_for_param(param_name: str, param_type: str) ->
     if param_type == 'int' or param_type == 'float':
         return '0'
     elif param_type == 'str':
-        return 'Random String'
+        return '"Random String"'
```
[the remark this answers is not in the document]
Comment (naterush) [reply mitosheet/dev/metaprogramming/utils/code_utils.py:30]: It's the value in Type script, so it needs to be in quotes.

File: mitosheet/mitosheet/tests/step_performers/test_excel_range_import.py (hunk 4 above)
Comment (naterush) [reply mitosheet/mitosheet/tests/step_performers/test_excel_range_import.py]: It literally won't let you write the same DF, unfortunately -- but all others are tests. Good tests - these caught some bugs!

**Pull request merged**